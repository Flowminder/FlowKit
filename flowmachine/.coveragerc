[run]
branch = True

[report]
fail_under = 90

omit =
    */python?.?/*
    */venv/*
    */site-packages/nose/*
    */tests/*
    */_version.py
    */versioneer.py
    *__init__*
    flowmachine/core/server/server.py

exclude_lines =
    if __name__ == '__main__':

show_missing = True
