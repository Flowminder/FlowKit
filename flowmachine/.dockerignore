build/
dist/
tests/
Dockerfile/
.ipynb_checkpoints/
.pytest_cache/
.flowmachine.egg-info/
docs/