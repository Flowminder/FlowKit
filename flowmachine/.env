# Set matplotlib backend to prevent ImportError when running in a virtual env
MPLBACKEND="agg"