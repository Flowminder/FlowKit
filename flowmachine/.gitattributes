flowmachine/_version.py export-subst
