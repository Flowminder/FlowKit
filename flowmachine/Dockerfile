# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

FROM python:3.7

COPY Pipfile* /app/
RUN pip install pipenv && cd app && pipenv install --deploy
COPY . /app/
WORKDIR /app
RUN pipenv run pip install .
CMD ["pipenv", "run", "flowmachine"]
ENV LOG_LEVEL=error
ENV DB_PORT=5432
ENV DB_HOST=flowdb
ENV DB_USER=analyst
ENV DB_PASS=flowflow
ENV DEBUG=False