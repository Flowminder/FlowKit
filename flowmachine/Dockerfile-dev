# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

FROM python:3.7

WORKDIR /app
COPY Pipfile* /app/
RUN pip install pipenv && pipenv install --dev
COPY . /app/
RUN pipenv install --skip-lock -e .
CMD pipenv run watchmedo auto-restart --recursive --patterns="*.py" --directory="." pipenv run flowmachine
ENV LOG_LEVEL=error
ENV FLOWDB_PORT=5432
ENV FLOWDB_HOST=flowdb
ENV FLOWDB_USER=analyst
ENV DB_PASS=flowflow
ENV DEBUG=False
