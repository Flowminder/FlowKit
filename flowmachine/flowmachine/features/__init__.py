# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

"""
Features (i.e. co-variates, indicators, metrics) are 
used as measurements of a phenomenon of interest.
This section of `flowmachine` contains code that calculates
a series of features.

"""
from .location import *
from .network import *
from .subscriber import *

from .raster import *
from .spatial import *

from .utilities import *

loc = [
    "TotalLocationEvents",
    "Flows",
    "UniqueSubscriberCounts",
    "LocationIntroversion",
    "MeaningfulLocationsAggregate",
    "MeaningfulLocationsOD",
]
nw = ["TotalNetworkObjects", "AggregateNetworkObjects"]
subs = [
    "RadiusOfGyration",
    "NocturnalEvents",
    "FirstLocation",
    "CallDays",
    "ModalLocation",
    "daily_location",
    "DayTrajectories",
    "LocationVisits",
    "NewSubscribers",
    "subscriber_location_cluster",
    "HartiganCluster",
    "UniqueLocationCounts",
    "SubscriberDegree",
    "TotalActivePeriodsSubscriber",
    "ContactBalance",
    "EventScore",
    "LabelEventScore",
    "SubscriberTACs",
    "SubscriberTAC",
    "SubscriberHandsets",
    "SubscriberHandset",
    "SubscriberHandsetCharacteristic",
    "ParetoInteractions",
    "SubscriberCallDurations",
    "PairedSubscriberCallDurations",
    "PerLocationSubscriberCallDurations",
    "PairedPerLocationSubscriberCallDurations",
    "MostFrequentLocation",
    "LastLocation",
    "PeriodicEntropy",
    "LocationEntropy",
    "ContactEntropy",
    "EventCount",
    "PerLocationEventStats",
    "PerContactEventStats",
    "MeaningfulLocations",
    "ProportionEventType",
    "PeriodicEntropy",
    "LocationEntropy",
    "ContactEntropy",
    "DistanceCounterparts",
    "ContactReferenceLocationStats",
    "IntereventPeriod",
    "TopUpAmount",
    "TopUpBalance",
    "MDSVolume",
    "ContactReciprocal",
    "ProportionContactReciprocal",
    "ProportionEventReciprocal",
    "HandsetStats",
    "VisitedMostDays",
]

rast = ["RasterStatistics"]
spat = [
    "LocationArea",
    "LocationCluster",
    "DistanceMatrix",
    "Geography",
    "VersionedInfrastructure",
    "Grid",
    "Circle",
    "CircleGeometries",
]

ut = [
    "GroupValues",
    "feature_collection",
    "SubscriberLocations",
    "EventTableSubset",
    "UniqueSubscribers",
    "EventsTablesUnion",
    "EventTableSubset",
]

sub_modules = ["location", "subscriber", "network", "utilities", "raster", "spatial"]

__all__ = loc + nw + subs + rast + ut + spat + sub_modules
