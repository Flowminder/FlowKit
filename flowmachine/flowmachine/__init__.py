# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

# -*- coding: utf-8 -*-
"""
FlowMachine is a Python based SQL wrapper built specifically for Flowminder's
PostgreSQL CDR database: Flowdb.

It provides a standardised way of interfacing with our postgres database. 
FlowMachine provides SQL based queries that compute commonly used features and
metrics to do with CDR data, such as `DailyLocations`, `ModalLocations`,
`Flows` and `RadiusOfGyration`.

The heart of flowmachine is the AbstractBaseClass Query, which defines how an
SQL-based feature behaves. All metrics define an SQL querty, 
and inherit from FlowMachine's main `Query()` class.

"""

from .versions import __version__
from .core.init import connect, connections
from .features.utilities import GroupValues, feature_collection
from flowmachine.core.logging import init_logging
import flowmachine.models
import flowmachine.features
import flowmachine.utils
import flowmachine.core

methods = ["GroupValues", "feature_collection", "connect", "connections"]
sub_modules = ["core", "features", "utils", "models"]
__all__ = methods + sub_modules

# Initialise loggers when flowmachine is imported
init_logging()

from ._version import get_versions

__version__ = get_versions()["version"]
del get_versions
