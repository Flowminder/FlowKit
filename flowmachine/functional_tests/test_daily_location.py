# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

from approvaltests.approvals import verify
from flowmachine.features import daily_location


def test_daily_location_1_sql(diff_reporter):
    """
    Simple daily location query returns the expected SQL string.
    """
    dl = daily_location("2016-01-01", "2016-01-02")
    sql = dl.get_query()
    verify(sql, diff_reporter)


def test_daily_location_1_df(get_dataframe, diff_reporter):
    """
    Simple daily location query returns the expected data.
    """
    dl = daily_location("2016-01-01", "2016-01-02")
    df = get_dataframe(dl)
    verify(df.to_csv(), diff_reporter)


def test_daily_location_2_sql(diff_reporter):
    """
    Daily location query with non-default parameters returns the expected SQL string.
    """
    dl = daily_location(
        "2016-01-04",
        level="admin2",
        hours=(3, 9),
        method="most-common",
        subscriber_identifier="imei",
        column_name="admin2pcod",
        ignore_nulls=False,
        subscriber_subset=["2GJxeNazvlgZbqj6", "7qKmzkeMbmk5nOa0", "8dpPLR15XwR7jQyN", "1NqnrAB9bRd597x2"],
    )
    sql = dl.get_query()
    verify(sql, diff_reporter)


def test_daily_location_2_df(get_dataframe, diff_reporter):
    """
    Daily location query with non-default parameters returns the expected data.
    """
    dl = daily_location(
        "2016-01-04",
        level="admin2",
        hours=(3, 9),
        method="most-common",
        #subscriber_identifier="imei",
        #column_name="admin2pcod",
        ignore_nulls=False,
        subscriber_subset=["2GJxeNazvlgZbqj6", "7qKmzkeMbmk5nOa0", "8dpPLR15XwR7jQyN", "1NqnrAB9bRd597x2"],
    )
    df = get_dataframe(dl)
    verify(df.to_csv(), diff_reporter)
