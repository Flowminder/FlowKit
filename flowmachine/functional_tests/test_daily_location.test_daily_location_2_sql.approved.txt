
        SELECT 
            ranked.subscriber, 
            admin2pcod
        FROM
             (SELECT times_visited.subscriber, admin2pcod,
             row_number() OVER (PARTITION BY times_visited.subscriber
              ORDER BY total DESC) AS rank
             FROM (
        SELECT 
            subscriber_locs.subscriber, 
            admin2pcod, 
            count(*) AS total
        FROM (
        SELECT
            l.subscriber, l.time, l.location_id,
            sites.admin2pcod
        FROM
            (
                SELECT
                    subscriber, datetime as time, location_id
                FROM
                    (
        SELECT datetime, imei AS subscriber, location_id
        FROM events.calls
        WHERE (datetime >= '2016-01-04'::timestamptz) AND (datetime <= '2016-01-05'::timestamptz)
         AND EXTRACT(hour FROM datetime) BETWEEN 3 and 8  AND  imei IN ('2GJxeNazvlgZbqj6', '7qKmzkeMbmk5nOa0', '8dpPLR15XwR7jQyN', '1NqnrAB9bRd597x2')
UNION ALL

        SELECT datetime, imei AS subscriber, location_id
        FROM events.sms
        WHERE (datetime >= '2016-01-04'::timestamptz) AND (datetime <= '2016-01-05'::timestamptz)
         AND EXTRACT(hour FROM datetime) BETWEEN 3 and 8  AND  imei IN ('2GJxeNazvlgZbqj6', '7qKmzkeMbmk5nOa0', '8dpPLR15XwR7jQyN', '1NqnrAB9bRd597x2')) AS foo
                
                ) AS l
        INNER JOIN
            (
        SELECT
            location_id, version, date_of_first_service, date_of_last_service,
            admin2pcod AS pcod
        FROM
            (
            SELECT
                locinfo.id AS location_id,
                locinfo.version,
                locinfo.date_of_first_service,
                locinfo.date_of_last_service,
                polygon.admin2pcod
            FROM
                infrastructure.cells AS locinfo
            INNER JOIN
                geography.admin2 AS polygon
            ON ST_within(
                locinfo.geom_point::geometry,
                ST_SetSRID(polygon.geom, 4326)::geometry
            )
            ) AS map
        ) AS sites
        ON
            l.location_id = sites.location_id
          AND
            l.time::date BETWEEN coalesce(sites.date_of_first_service,
                                                '-infinity'::timestamptz) AND
                                       coalesce(sites.date_of_last_service,
                                                'infinity'::timestamptz)
         ORDER BY time) AS subscriber_locs
        GROUP BY subscriber_locs.subscriber, admin2pcod
        ) AS times_visited) AS ranked
        WHERE rank = 1
        